### DISCUSSION TITLE
Is my env variable save?

### DISCUSSION THREAD

**Discussion by md186** [vercel/next.js#26829] [Help]
Hello,

I have several api endpoints which are fetching data with get or create data in my database and send mails with post.

i want to protect my api routes so that if the api is called only a few domains and the application itself can access the endpoints.

i thought about the following concept: Im saving the domains which are allowed to bypass with tokens in my database and if the api is called I’m checking if the token is valid. This will work perfectly fine I guess.

im only struggling with the case if the own application is doing such a request. From the official docs I learned the env variables are secret inside getinitialprops serversideprops and on the api route.

Now imagine we have a button on the clientside and if you press it my function will be called which does a post on my endpoint. The endpoint awaits the domain and a token, now if I would place the env var (which contains the valid token for the own application) in the request there , it would be undefined since on the client the var isn’t readable. But what if we create a var inside the getinitialprops and save the env value and return it as props to take that for the request later on?

The var should be readable than but I’m not sure if the value is still secure or not. Can somebody confirm if this is still secure or not and a valid approach?

**Comment by robertpiosik** [comment]
Well, you will have to embrace client side authentication. Anything bundled within your app is publicly visible.

**Reply by md186** [reply]
But if I save the env var in a var inside getinitialprops and pass it to my request on the clientside it shouldn’t be public or?

**Reply by md186** [reply]
Im passing it as props there

**Reply by robertpiosik** [reply]
If you know the user requesting the page, he is authenticated then. So you could simply send these requests for protected routes with the same cookie and authenticate it again on a route guard.

**Reply by md186** [reply]
I have the following cases: I have two  external pages and I will give them tokens which are saved in my database so they can fetch the APIs. (For example posting given data to create a database entry)

im only struggling the case if a User on my page is doing a post within a form. In my App their is no authentification needed they should be able to do the Post via a form also, but I need to make sure that it’s coming from my App. Thats why.

how to do this? That’s why I thought I’m saving my own token for the app in an env var and pass it threw getinitlaprops so I can confirm it’s requesting from my app

**Reply by md186** [reply]
The database creation does only work on api endpoints or inside getinitialprops thats why I’m thinking about that way. But I can’t call a function on getinitialprops later on if a user fills in a form right ?
